### DISCUSSION TITLE
Pyinstaller drag-and-drop files to the --windowed app

### DISCUSSION THREAD

**Discussion by Moosems** [pyinstaller/pyinstaller#6828] [Help] [answered]
# Pyinstaller drag-and-drop files to the --windowed app

How do I allow files to be dragged and dropped onto the --windowed app and allow for it to be seen as a text editor/code editor by the system (Be recommended when you click open with and can have files dragged onto it)? [This question from stackoverflow](https://stackoverflow.com/questions/16712000/pyinstaller-drag-and-drop-files-to-the-onefile-exe) is practically the same to mine but since I am using mac it seems that the answer is different. I could not get the suggested answer to work. When I try to drag files onto the app it refuses to open it and doesn't come up as recommended

**Comment by bwoodsend** [comment]
That's all documented [here](https://pyinstaller.org/en/v5.0.1/feature-notes.html?highlight=drag%20#macos-event-forwarding-and-argv-emulation-in-app-bundles).

**Reply by Moosems** [reply]
How do I have it come up in `Recommended Applications`?

**Reply by Moosems** [reply]
@bwoodsend  Issue: Sometimes it won't let me drag files onto the app even though I used --argv-emulation. Why is that? It worked for a while and then didn't all of a sudden. Not sure how to reproduce as I can't get it to work at all now

My command line instruction:
`
pyinstaller JDE.py --add-data "settings.txt:." --add-data "keywords.txt:." --add-data "run.png:." --add-data "clear_terminal.png:." --add-data "clear_text.png:." --add-data "color_mode.png:."  --add-data "copy.png:." --add-data "create_template.png:." --add-data "delete_template.png:." --add-data "find_and_replace.png:." --add-data "open_template.png:." --add-data "open.png:." --add-data "paste.png:." --add-data "quit.png:." --add-data "report_bug.png:." --add-data "save_as.png:." --add-data "save.png:." --add-data "settings.png:." --add-data "reset.png:." --add-data "export_template.png:." --add-data "extensions.png:." --add-data "color_theme.txt:." --add-data "font.txt:." --icon=JDE.icns --windowed --argv-emulation
`

Image showing error:
<img width="1440" alt="Screen Shot 2022-05-17 at 7 25 51 PM" src="https://user-images.githubusercontent.com/95927277/168928257-eaabd350-4554-448b-bbf5-42a3637d745b.png">

NOTE:

I kept doing that over and over and sometimes it did work so it seems to almost be at random

**Comment by bwoodsend** [comment, marked as answer]
> Sometimes it won't let me drag files onto the app even though I used --argv-emulation.

`--argv-emulation` has nothing to do with drag and drop.

**Reply by Moosems** [reply]
What do you mean? That's what allowed me to drag files onto it before. What should I be doing then?

**Reply by rokm** [reply]
> What should I be doing then?

You should read the whole section on macOS event forwarding and argv emulation.

Argv emulation only lets you handle the initial open (drag) event(s). Once the application is running, you need to handle any subsequent events yourself, by installing event handlers in the UI toolkit that you are using.

**Reply by Moosems** [reply]
@rokm I understand but it won't even do the initial handling. I've read the section 5 times over (In the most literal sense)  and still can't get it to let files be dragged onto it. I don't know what to do. It handles everything after that (works when you force it to open the file with `Open With`,  `Other`,  `Enable: All Applications`) but it won't allow files to drag onto it

**Reply by rokm** [reply]
Does it disallow files to be dragged onto the bundle, or does it show an error dialog?

**Reply by Moosems** [reply]
It disallows it

**Reply by rokm** [reply]
Have you registered the supported extensions in the Info.plist?

I tried running the Tk-based event logger from the docs, and dragging seems to work for file types that are declared as supported. But the very first attempt at open (until the app is running) seems to fail, which is something I need to look into. EDIT: seems to be a "feature" of Tcl/Tk 8.6.11 bundled with my python 3.9.4 installation. It works as expected with python 3.10.2 and Tcl/Tk 8.6.12.

**Reply by Moosems** [reply]
@rokm How do I add it to the Info.plist? Can that be done from the command line?

**Reply by rokm** [reply]
No, you need to create a .spec file, edit it, and then build from .spec instead directly from .py file. 

Or, you may be able edit the file afterwards with a text editor, but you'll probably need to re-sign the bundle afterwards.

**Reply by rokm** [reply]
See [here](https://pyinstaller.org/en/v5.0.1/feature-notes.html?#registering-supported-file-types-and-custom-url-schemas) or the examples further down in the section.

**Reply by Moosems** [reply]
@rokm How would I add folders as supported? I'm reading through the examples but am still a little confused

**Reply by rokm** [reply]
When you run `pyinstaller ... program.py`, it generates a `program.spec` file, and then builds the program using that. So you can take that .spec file as a base and edit it, then build your program by running `pyinstaller program.spec`.

And in the spec file, you edit the `BUNDLE()` arguments to include `info_plist` dictionary with relevant info.

**Reply by Moosems** [reply]
> When you run `pyinstaller ... program.py`, it generates a `program.spec` file, and then builds the program using that. So you can take that .spec file as a base and edit it, then build your program by running `pyinstaller program.spec`.
> 
> And in the spec file, you edit the `BUNDLE()` arguments to include `info_plist` dictionary with relevant info.

@rokm I figured that out but adding the args to the dictionary is confusing me. I figured out the python files and txt files but can't figure out dir

**Reply by rokm** [reply]
Hmm, I'm not sure about folders, either. [This](https://github.com/microsoft/vscode/issues/146977) gives an [example](https://github.com/ImageOptim/ImageOptim/blob/d8d6bc528f0b19253585d824851e320db3261d03/imageoptim/Info.plist#L107-L118). So I suppose after remapping dicts and arrays to our spec format:

```
info_plist={
        'CFBundleDocumentTypes': [
        {
            'CFBundleTypeName': 'Python files',
            'CFBundleTypeExtensions': [
                'py'
            ],
            'CFBundleTypeRole': "Editor",
         },
         {
            'CFBundleTypeName': 'Folder',
            'CFBundleTypeRole': "Editor",
            'LSItemContentTypes': ['public.folder']
         }
         ],
    }

```

**Reply by Moosems** [reply]
@rokm Can you add that as the answer so I can mark it as the answer?

**Reply by rokm** [reply]
Just mark this thread as the answer. It will probably be more helpful to anyone else who comes across it, as it contains the whole information.

**Reply by Moosems** [reply]
@rokm Is there any way to make it only allow files at start?

**Reply by rokm** [reply]
> @rokm Is there any way to make it only allow files at start?

You probably cannot prevent files from being dragged onto the app bundke, because that's a static property.

**Reply by Moosems** [reply]
@rokm Are there any bindings/events I can use to detect further dragging when using tkinter?

**Reply by rokm** [reply]
[Yes](https://pyinstaller.org/en/v5.1/feature-notes.html#open-event-handling-in-a-tkinter-based-gui-application). But as far as I can tell, they don't allow you to reject the event. So for your purposes, it's probably the same as not having them registered in the first place.
